app/src/org/commcare/sync/FirebaseMessagingDataSyncer.java

```diff
@@ -0,0 +1,334 @@
+        boolean recreateActivity = false;
+
+        try {
+            switch (getAppNavigationState()) {
+                case ENTITY_SELECTION:
+                    aSessWrapper.cleanVolatiles();
+                    ((EntitySelectActivity) activity).loadEntities();
+                    break;
+                case ENTITY_DETAIL:
+                    aSessWrapper.cleanVolatiles();
+                    recreateActivity = true;
+                    break;
+                case MENU:
+                    recreateActivity = true;
+                    break;
+            }
+            commcareSession.syncState();
+            if (recreateActivity) {
+                activity.recreate();
+            }
+        }
+        catch(RuntimeException e){
+            activity.setResult(RESULT_CANCELED);
+            activity.finish();
+        }
+    }
+
+    private boolean checkUserAndDomain(User user, String payloadUsername, String payloadDomain) {
+        if(payloadUsername != null && payloadDomain != null){
+            String loggedInUsername = user.getUsername();
+            String userDomain = removeServerUrlFromUserDomain(HiddenPreferences.getUserDomain());
+            return payloadUsername.equalsIgnoreCase(loggedInUsername) && payloadDomain.equalsIgnoreCase(userDomain);
+        }
+        return false;
+    }
+
+    private AppNavigationStates getAppNavigationState() {
+        if (CommCareActivity.currentActivity instanceof FormEntryActivity)
+            return FORM_ENTRY;
+        else if (CommCareActivity.currentActivity instanceof EntitySelectActivity)
+            return ENTITY_SELECTION;
+        else if (CommCareActivity.currentActivity instanceof EntityDetailActivity)
+            return ENTITY_DETAIL;
+        else if (CommCareActivity.currentActivity instanceof StandardHomeActivity)
+            return HOME_SCREEN;
+        else if (CommCareActivity.currentActivity instanceof MenuActivity)
+            return MENU;
+        return OTHER;
+    }
+
+    // This method is responsible for informing the User and scheduling a sync for when it's safe
+    private void informUserAboutPendingSync(CommCareActivity activity, FCMMessageData fcmMessageData) {
+        activity.runOnUiThread(() ->
+                activity.alertPendingSync(fcmMessageData)
+        );
+    }
+
+    @Override
+    public void connectTask(CommCareTask task) {
+        this.currentTask = task;
+    }
+
+    @Override
+    public void startBlockingForTask(int id) {
+        // In case a background sync is already ongoing, we ignore the new request silently
+        if (CommCareSessionService.sessionAliveLock.isLocked()) {
+            currentTask.cancel(true);
+        }
+        mPinnedNotificationProgress = new PinnedNotificationWithProgress(context,
+                "sync.communicating.title","sync.progress.starting", -1);
+
+        // Disable any pending sync
+        disablePendingSyncs();
```

Comment:
I moved this here on purpose, the rationale being we are supposed to receive several sync requests in short periods of time, which means that these requests are supposed to be short-lived. So the moment we trigger one, we just release any pending sync related flags. But logically I can understand that it can be confusing, I will move this back.

Reply:
we would not want to trigger simultaneous syncs anyway so we should ensure that 2 push notifications back to back only initiates sync request once.